sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -74,6 +66,160 @@ object TableOutputResolver {
+    val fields = inputType.zipWithIndex.map { case (f, i) =>
+      Alias(GetStructField(input, i, Some(f.name)), f.name)()
+    }
+    val reordered = reorderColumnsByName(fields, expectedType.toAttributes, conf, addError, colPath)
+    if (reordered.length == expectedType.length) {
+      val struct = CreateStruct(reordered)
+      val res = if (input.nullable) {
+        If(IsNull(input), Literal(null, struct.dataType), struct)
+      } else {
+        struct
+      }
+      Some(Alias(res, expectedName)())
+    } else {
+      None
+    }
+  }
+
+  private def resolveArrayType(
+      input: NamedExpression,
+      inputType: ArrayType,
+      expectedType: ArrayType,
+      expectedName: String,
+      conf: SQLConf,
+      addError: String => Unit,
+      colPath: Seq[String]): Option[NamedExpression] = {
+    if (inputType.containsNull && !expectedType.containsNull) {
+      addError(s"Cannot write nullable elements to array of non-nulls: '${colPath.quoted}'")
+      None
+    } else {
+      val param = NamedLambdaVariable("x", inputType.elementType, inputType.containsNull)
+      val fakeAttr = AttributeReference("x", expectedType.elementType, expectedType.containsNull)()
+      val res = reorderColumnsByName(Seq(param), Seq(fakeAttr), conf, addError, colPath)
+      if (res.length == 1) {
+        val func = LambdaFunction(res.head, Seq(param))
+        Some(Alias(ArrayTransform(input, func), expectedName)())
+      } else {
+        None
+      }
+    }
+  }
+
+  private def resolveMapType(
+      input: NamedExpression,
+      inputType: MapType,
+      expectedType: MapType,
+      expectedName: String,
+      conf: SQLConf,
+      addError: String => Unit,
+      colPath: Seq[String]): Option[NamedExpression] = {
+    if (inputType.valueContainsNull && !expectedType.valueContainsNull) {
+      addError(s"Cannot write nullable values to map of non-nulls: '${colPath.quoted}'")
+      None
+    } else {
+      val keyParam = NamedLambdaVariable("k", inputType.keyType, nullable = false)
+      val fakeKeyAttr = AttributeReference("k", expectedType.keyType, nullable = false)()
+      val resKey = reorderColumnsByName(
+        Seq(keyParam), Seq(fakeKeyAttr), conf, addError, colPath :+ "key")
+
+      val valueParam = NamedLambdaVariable("v", inputType.valueType, inputType.valueContainsNull)
+      val fakeValueAttr =
+        AttributeReference("v", expectedType.valueType, expectedType.valueContainsNull)()
+      val resValue = reorderColumnsByName(
+        Seq(valueParam), Seq(fakeValueAttr), conf, addError, colPath :+ "value")
+
+      if (resKey.length == 1 && resValue.length == 1) {
+        val keyFunc = LambdaFunction(resKey.head, Seq(keyParam))
+        val valueFunc = LambdaFunction(resValue.head, Seq(valueParam))
+        val newKeys = ArrayTransform(MapKeys(input), keyFunc)
+        val newValues = ArrayTransform(MapValues(input), valueFunc)
+        Some(Alias(MapFromArrays(newKeys, newValues), expectedName)())
```

Comment:
`TransformValues(TransformKeys(input, keyFunc), valueFunc)`?

Reply:
This creates map twice, can be slower.

Reply:
ok